src/problems/cvrp/operators/or_opt.cpp

```diff
@@ -40,88 +41,26 @@ OrOpt::OrOpt(const Input& input,
-      // Adding edge past the end after a real job.
-      auto p_index = _input.jobs[t_route[t_rank - 1]].index();
-      previous_cost = t_v.eval(p_index, s_index);
-      reverse_previous_cost = t_v.eval(p_index, after_s_index);
-      if (t_v.has_end()) {
-        auto n_index = t_v.end.value().index();
-        old_edge_cost = t_v.eval(p_index, n_index);
-        next_cost = t_v.eval(after_s_index, n_index);
-        reverse_next_cost = t_v.eval(s_index, n_index);
-      }
-    }
-  } else {
-    // Adding before one of the jobs.
-    auto n_index = _input.jobs[t_route[t_rank]].index();
-    next_cost = t_v.eval(after_s_index, n_index);
-    reverse_next_cost = t_v.eval(s_index, n_index);
-
-    if (t_rank == 0) {
-      if (t_v.has_start()) {
-        auto p_index = t_v.start.value().index();
-        previous_cost = t_v.eval(p_index, s_index);
-        reverse_previous_cost = t_v.eval(p_index, after_s_index);
-        old_edge_cost = t_v.eval(p_index, n_index);
-      }
-    } else {
-      auto p_index = _input.jobs[t_route[t_rank - 1]].index();
-      previous_cost = t_v.eval(p_index, s_index);
-      reverse_previous_cost = t_v.eval(p_index, after_s_index);
-      old_edge_cost = t_v.eval(p_index, n_index);
-    }
-  }
-
-  // Gain for source vehicle, including cost of moved edge.
-  s_gain =
-    _sol_state.edge_gains[s_vehicle][s_rank] + s_v.eval(s_index, after_s_index);
-
-  if (s_route.size() == 2) {
-    s_gain.cost += s_v.fixed_cost();
-  }
-
-  // Gain for target vehicle, including cost of moved edge.
-  _normal_t_gain = old_edge_cost - previous_cost - next_cost -
-                   t_v.eval(s_index, after_s_index);
-
-  _reversed_t_gain = old_edge_cost - reverse_previous_cost - reverse_next_cost -
-                     t_v.eval(after_s_index, s_index);
-
-  if (t_route.empty()) {
-    _normal_t_gain.cost -= t_v.fixed_cost();
-    _reversed_t_gain.cost -= t_v.fixed_cost();
-  }
+  s_gain = utils::removal_cost_delta(_input, _sol_state, source, s_rank, 2);
+
+  _normal_t_gain = utils::addition_cost_delta(_input,
+                                              _sol_state,
+                                              target,
+                                              t_rank,
+                                              t_rank,
+                                              source,
+                                              s_rank,
+                                              s_rank + 2);
+
+  _reversed_t_gain = utils::addition_cost_delta(_input,
+                                                _sol_state,
+                                                target,
+                                                t_rank,
+                                                t_rank,
+                                                source,
+                                                s_rank,
+                                                s_rank + 2,
+                                                REVERSED_INSERTION);
```

Comment:
This is typical of all operators potentially reversing some edges (`OrOpt`, `{Mixed,Cross}Exchange` and the `Intra*` versions): we now have two calls to `addition_cost_delta` but basically those two call we mostly perform the same calculation except for the small portion evaluating new inserted cost, either reversed or not.

This is less efficient than the previous implementation where everything was evaluated "manually" here, but both cost versions where computed in one go.